### DISCUSSION TITLE
Dificuldade com composição de React hooks com TypeScript

### DISCUSSION THREAD

**Discussion by tassoevan** [frontendbr/forum#1813] [Perguntas e Respostas]
Olá, tudo bem?

Estou fazendo a transição de um pequeno conjunto de React hooks de JavaScript para TypeScript. Um destes hooks, [`useSafely()`](https://github.com/RocketChat/Rocket.Chat.Fuselage/blob/develop/packages/fuselage-hooks/src/useSafely.js), serve para composição com os hooks primitivos de estado `useReducer()` e `useState()`:

```ts
const [value, setValueSafely] = useSafely(useState(initialValue));
const [state, safeDispatch] = useSafely(useReducer(reducer, initialState));
```

Basicamente ele faz o "trabalho sujo" de evitar uma mutação acidental do estado de um componente já desmontado (futuramente espero evitar memory leaks também), algo que infelizmente é necessário enquanto procuramos uma solução mais robusta para lidar com side effects assíncronos.

O problema em específico que estou encontrando se refere a declarar [uma versão tipada de `useSafely()`](https://github.com/RocketChat/Rocket.Chat.Fuselage/blob/89ac92c55ab8eeb72dca42f924ef7ab07034b6a1/packages/fuselage-hooks/src/useSafely.ts). O seu único argumento é um par do tipo `[S, React.DispatchWithoutAction | React.Dispatch<A>]`, onde `S` é um tipo genérico para o estado e `A` para ação enviada por `dispatch`. Em `useState<S>()` o retorno é um par do tipo `[S, React.Dispatch<React.SetStateAction<S>>]` mas `useReducer()` pode prover tanto `[S, React.Dispatch<A>]` e `[S, React.DispatchWithoutAction]`. De alguma forma essa ambiguidade tem causado um problema de inferência de tipo do segundo elemento do par com o modo estrito ativado (`"strict": true` no `tsconfig.json`). A função `dispatch` pode ser tanto `() => void` quando `(action: A) => void` mas por algum motivo do `tsc` entende que isto

```ts
const safeDispatch = useCallback((action?: A) => { // usamos um hook similar a useCallback
    if (!mountedRef.current) {
      return;
    }

    dispatch(action);
  }, []);
```

só pode representar que `safeDispatch` é `React.Dispatch<A>`.

O que estou fazendo de errado?

**Comment by ninetails** [comment]
Sei que não é de ajuda mas tem como portar um exemplo onde quebre usando sem o `useSafely` num CodeSandbox?

Diria pra tentar ajudar, tentou colocar alguma validação onde jogue um erro quando não for `React.DispatchWithoutAction`, assim aceitando a união? Talvez uma mensagem de que não suporta no momento... Desfazer a união talvez seria mais complicado que restringir, pensando no caso de tratar todos os casos...
